readme.md
## GitDeclare

** This is a personal tool. I've only tested it on my own system. You're more than welcome to use the program or code, but there is no official documentation. **

Declarative Git Processing. Instead of committing haphazardly, GitDeclare watches for all changes every second and auto commits them for you. You can declare your intentions to work on something and then reap the benefits of having a snapshot of all changes done in a session.

In addition, if you don't like the changes, you can delete them.

Changelog:
#### 24/02/2018: 18:53 - 18:56:pool[fix timestamp for showing the time worked for each commit]
#### 24/02/2018: 18:53 - 18:57:pool[testing]
#### 24/02/2018: 18:59 - 18:59:pool[testing!]
#### 24/02/2018: 18:59 - 18:59:pool[testing again]
#### 24/02/2018: 19:00 - 19:01:pool[testing for long periods of time and multiple commits]
#### 24/02/2018: 19:01 - 19:02:pool[still testing]
#### 25/02/2018: 09:32 - 09:51:pool[change formatting of markdown generated for readme]
#### 25/02/2018: 09:51 - 10:12:pool[work on new option to switch branch]
#### 25/02/2018: 10:12 - 10:12:pool[testing]
##### 25/02/2018: 10:16 - 10:16:pool[testing switch feature]
##### 25/02/2018: 10:16 - 10:16:pool[testing switch]
##### 25/02/2018: 10:16 - 10:17:pool[finalize testing on switch feature]
##### 25/02/2018: 10:17 - 10:19:pool[change formatting of exit options]
##### 25/02/2018: 10:25 - 10:25:pool[test]

##### 31/03/2018: 18:30 - 18:32:pool[change readme and changelog output from file directory to current working directory]

##### 31/03/2018: 18:32 - 18:37:pool[add ability to execute remotely]

##### 31/03/2018: 18:49 - 19:13:pool[testing]

##### 02/04/2018: 19:14 - 19:34:pool[begin creation of gitdeclare http]

##### 02/04/2018: 19:34 - 19:35:pool[small change to start method to show current branch name and skip asking what branch is being worked on]

##### 02/04/2018: 19:55 - 19:55:pool[test]

##### 02/04/2018: 19:55 - 19:55:pool[test2]

##### 02/04/2018: 19:57 - 19:57:pool[testing]

##### 02/04/2018: 19:58 - 19:59:pool[another test]

##### 02/04/2018: 19:35 - 20:00:pool[testing httparty pushing with sandbox]

##### 03/04/2018: 19:36 - 19:36:pool[testing]

##### 03/04/2018: 19:37 - 19:37:pool[testing 2]

##### 03/04/2018: 19:39 - 19:40:pool[testing again]

##### 03/04/2018: 19:40 - 19:41:pool[another test]

##### 03/04/2018: 19:22 - 19:48:pool[finish testing sandbox posting to test api]

##### 03/04/2018: 19:48 - 19:52:pool[begin migration of sandbox changes to gdec_http]

##### 03/04/2018: 19:52 - 19:55:pool[migrate fix to start method to gitdeclare.rb and gdec cli]

##### 03/04/2018: 19:55 - 19:59:pool[create httpdec cli]

##### 07/04/2018: 16:56 - 16:57:pool[test 1]

##### 07/04/2018: 16:57 - 16:57:pool[test 2]

##### 07/04/2018: 16:57 - 16:57:pool[test 3]

##### 08/04/2018: 17:15 - 17:25:pool[implement working directory and git uri in post to declarations]

##### 08/04/2018: 17:25 - 17:28:pool[migrate changes from sandbox to production]

##### 08/04/2018: 17:35 - 17:36:pool[testing 1]

##### 08/04/2018: 17:37 - 17:38:pool[testing POST]


changelog.txt
24/02/2018: 18:53 - 18:56:pool[fix timestamp for showing the time worked for each commit]
24/02/2018: 18:53 - 18:57:pool[testing]
24/02/2018: 18:59 - 18:59:pool[testing!]
24/02/2018: 18:59 - 18:59:pool[testing again]
24/02/2018: 19:00 - 19:01:pool[testing for long periods of time and multiple commits]
24/02/2018: 19:01 - 19:02:pool[still testing]
25/02/2018: 09:32 - 09:51:pool[change formatting of markdown generated for readme]
25/02/2018: 09:51 - 10:12:pool[work on new option to switch branch]
25/02/2018: 10:12 - 10:12:pool[testing]
25/02/2018: 10:16 - 10:16:pool[testing switch feature]
25/02/2018: 10:16 - 10:16:pool[testing switch]
25/02/2018: 10:16 - 10:17:pool[finalize testing on switch feature]
25/02/2018: 10:17 - 10:19:pool[change formatting of exit options]
25/02/2018: 10:25 - 10:25:pool[test]
31/03/2018: 18:30 - 18:32:pool[change readme and changelog output from file directory to current working directory]
31/03/2018: 18:32 - 18:37:pool[add ability to execute remotely]
31/03/2018: 18:49 - 19:13:pool[testing]
02/04/2018: 19:14 - 19:34:pool[begin creation of gitdeclare http]
02/04/2018: 19:34 - 19:35:pool[small change to start method to show current branch name and skip asking what branch is being worked on]
02/04/2018: 19:55 - 19:55:pool[test]
02/04/2018: 19:55 - 19:55:pool[test2]
02/04/2018: 19:57 - 19:57:pool[testing]
02/04/2018: 19:58 - 19:59:pool[another test]
02/04/2018: 19:35 - 20:00:pool[testing httparty pushing with sandbox]
03/04/2018: 19:36 - 19:36:pool[testing]
03/04/2018: 19:37 - 19:37:pool[testing 2]
03/04/2018: 19:39 - 19:40:pool[testing again]
03/04/2018: 19:40 - 19:41:pool[another test]
03/04/2018: 19:22 - 19:48:pool[finish testing sandbox posting to test api]
03/04/2018: 19:48 - 19:52:pool[begin migration of sandbox changes to gdec_http]
03/04/2018: 19:52 - 19:55:pool[migrate fix to start method to gitdeclare.rb and gdec cli]
03/04/2018: 19:55 - 19:59:pool[create httpdec cli]
07/04/2018: 16:56 - 16:57:pool[test 1]
07/04/2018: 16:57 - 16:57:pool[test 2]
07/04/2018: 16:57 - 16:57:pool[test 3]
08/04/2018: 17:15 - 17:25:pool[implement working directory and git uri in post to declarations]
08/04/2018: 17:25 - 17:28:pool[migrate changes from sandbox to production]
08/04/2018: 17:35 - 17:36:pool[testing 1]
08/04/2018: 17:37 - 17:38:pool[testing POST]
01/05/2018: 19:52 - 19:53:pool[implement a next goal feature at the end of a push to declare future changes]


command_test.rb
x = %x(git rev-parse --abbrev-ref HEAD)
branch = x
puts "On #{branch} branch"

gdec.rb



class GitDeclare
    @@pushes = 0
    @@stage = 0
    @@changes = []
    @@color_red = "\033[31m"
    @@color_green = "\033[32m"
    @@color_default = "\033[0m"
    @@commits = 1
    @@starttime = Time.now.strftime("%H:%M")
    @@date = Time.now.strftime("%d/%m/%Y")
    @@time = nil
    @@pool = nil
    @@branch = nil

    def initialize; end

    def self.current_time
        Time.now.strftime("%H:%M")
    end

    def self.execute(param)
        stalker = %x{#{param}}
        
        if stalker.include? "nothing to commit" 
            
        elsif stalker.include? "insert"
            puts @@color_green + stalker + @@color_default
            puts "#{@@commits} changes saved"
            @@commits += 1
        end
    end

    def self.add_wait
        sleep 1
        GitDeclare.execute "git add ."
        sleep 1
    end

    def self.commit_loop(pool)
            GitDeclare.add_wait
            GitDeclare.execute "git commit -m \" #{pool} \""
    end

    def self.atomic(summary, pool)
        open('changelog.txt', 'a') do |file|
            file.puts "#{@@date}: #{@@time} - #{GitDeclare.current_time}:pool[#{pool}]"
        end
        open('readme.md', 'a') do |file|
            file.puts "\n##### #{@@date}: #{@@time} - #{GitDeclare.current_time}:pool[#{pool}]"
        end
        
        @@changes << pool
        if @@stage == 1
            @@changes.map! {|item| item = "* #{item.strip}"}
            open('pull_me.txt', 'a') do |file|
                file.puts "[#{summary}]"
                file.puts "### #{@@date}[#{@@starttime} - #{GitDeclare.current_time}]:"
                file.puts @@changes
                file.puts
            end
        end
        GitDeclare.add_wait
        GitDeclare.execute "git commit -m \"#{pool}\""
        
    end

    def self.exit(exit_type, pool, branch)
        case exit_type
        when "new"
            GitDeclare.atomic(nil, pool)
            GitDeclare.start
        when "reset"
            puts "Wiping commits and exiting"
            system "git reset HEAD~"
        when "push"
            puts "Summarize final changes:"
            summary = gets.chomp
            @@stage = 1
            GitDeclare.atomic(summary, pool)
            GitDeclare.execute "git push -u origin #{branch}"
        when "switch"
            GitDeclare.atomic(nil, pool)
            puts "What branch are you working on?"; @@branch = gets.chomp
            GitDeclare.start
        else
            puts "Returning to loop"
            GitDeclare.threader(branch)
        end
    end

    def self.threader(branch)
        puts "What are you working on with the #{branch} branch?"
        @@pool = gets.chomp
        puts "You're now working on: \"#{@@pool}\" on #{branch} branch. #{@@color_red}GitDeclare is watching for changes#{@@color_default}."
        puts "When you're done with this change, press [Enter] to make a commit and start a new declaration."
        declare = Thread.new do
            
            while true
                GitDeclare.commit_loop(@@pool)
            end
            
        end
        
        gets
        declare.kill
        puts "How do you wish to exit?"
        puts "
        'new': logs the commit pool and starts a new declaration
        'reset': wipes commits and exits program
        'push': pushes all changes
        'switch': logs commit, starts new declare, and switches branch"
        exit_type = gets.chomp
        GitDeclare.exit(exit_type, @@pool, branch)
        
        
    end

    def self.start
        @@time = GitDeclare.current_time
        @@pushes > 0 ? @@pushes += 1 : open('pull_me.txt', 'w') {|f| f.puts ""}; @@pushes += 1
        if @@branch == nil then puts "What branch are you working on?"; @@branch = gets.chomp end
        
        GitDeclare.threader(@@branch)
    end

    

end

GitDeclare.start


gdec.1.rb
#!/usr/bin/env node


class GitDeclare
    @@pushes = 0
    @@stage = 0
    @@changes = []
    @@color_red = "\033[31m"
    @@color_green = "\033[32m"
    @@color_default = "\033[0m"
    @@commits = 1
    @@starttime = Time.now.strftime("%H:%M")
    @@date = Time.now.strftime("%d/%m/%Y")
    @@time = nil
    @@pool = nil
    @@branch = nil

    def initialize; end

    def self.current_time
        Time.now.strftime("%H:%M")
    end

    def self.execute(param)
        stalker = %x{#{param}}
        
        if stalker.include? "nothing to commit" 
            
        elsif stalker.include? "insert"
            puts @@color_green + stalker + @@color_default
            puts "#{@@commits} changes saved"
            @@commits += 1
        end
    end

    def self.add_wait
        sleep 1
        GitDeclare.execute "git add ."
        sleep 1
    end

    def self.commit_loop(pool)
            GitDeclare.add_wait
            GitDeclare.execute "git commit -m \" #{pool} \""
    end

    def self.atomic(summary, pool)
        open("#{Dir.pwd}/changelog.txt", 'a') do |file|
            file.puts "#{@@date}: #{@@time} - #{GitDeclare.current_time}:pool[#{pool}]"
        end
        open("#{Dir.pwd}/readme.md", 'a') do |file|
            file.puts "\n##### #{@@date}: #{@@time} - #{GitDeclare.current_time}:pool[#{pool}]"
        end
        
        @@changes << pool
        if @@stage == 1
            @@changes.map! {|item| item = "* #{item.strip}"}
            open('pull_me.txt', 'a') do |file|
                file.puts "[#{summary}]"
                file.puts "### #{@@date}[#{@@starttime} - #{GitDeclare.current_time}]:"
                file.puts @@changes
                file.puts
            end
        end
        GitDeclare.add_wait
        GitDeclare.execute "git commit -m \"#{pool}\""
        
    end

    def self.exit(exit_type, pool, branch)
        case exit_type
        when "new"
            GitDeclare.atomic(nil, pool)
            GitDeclare.start
        when "reset"
            puts "Wiping commits and exiting"
            system "git reset HEAD~"
        when "push"
            puts "Summarize final changes:"
            summary = gets.chomp
            @@stage = 1
            GitDeclare.atomic(summary, pool)
            GitDeclare.execute "git push -u origin #{branch}"
        when "switch"
            GitDeclare.atomic(nil, pool)
            puts "What branch are you working on?"; @@branch = gets.chomp
            GitDeclare.start
        else
            puts "Returning to loop"
            GitDeclare.threader(branch)
        end
    end

    def self.threader(branch)
        puts "What are you working on with the #{branch} branch?"
        @@pool = gets.chomp
        puts "You're now working on: \"#{@@pool}\" on #{branch} branch. #{@@color_red}GitDeclare is watching for changes#{@@color_default}."
        puts "When you're done with this change, press [Enter] to make a commit and start a new declaration."
        declare = Thread.new do
            
            while true
                GitDeclare.commit_loop(@@pool)
            end
            
        end
        
        gets
        declare.kill
        puts "How do you wish to exit?"
        puts "
        'new': logs the commit pool and starts a new declaration
        'reset': wipes commits and exits program
        'push': pushes all changes
        'switch': logs commit, starts new declare, and switches branch"
        exit_type = gets.chomp
        GitDeclare.exit(exit_type, @@pool, branch)
        
        
    end

    def self.start
        @@time = GitDeclare.current_time
        @@pushes > 0 ? @@pushes += 1 : open('pull_me.txt', 'w') {|f| f.puts ""}; @@pushes += 1
        if @@branch == nil then puts "What branch are you working on?"; @@branch = gets.chomp end
        
        GitDeclare.threader(@@branch)
    end

    

end

GitDeclare.start


gr_net.rb
=begin

    DEPRECATED

=end

require 'http_require'
require 'http://www.viktharienvolander.com/threader.rb'

class GitReaper
    include Threader
    @@pushes = 0
    @@color_red = "\033[31m"
    @@color_green = "\033[32m"
    @@color_default = "\033[0m"
    @@commits = 1
    @@time_running = 0

    def initialize
        
    end

    def self.execute(param)
        stalker = %x{#{param}}
        @@time_running += 1
        if stalker.include? "nothing to commit" 
            puts @@color_red + "Stalking for #{@@time_running} secs" + @@color_default
        elsif stalker.include? "insert"
            puts @@color_green + stalker + @@color_default
            puts "#{@@commits} commits to pool so far"
            @@commits += 1
        end
    end

    def self.add_wait
        sleep 1
        GitReaper.execute "git add ."
        sleep 1
    end

    def self.commit_loop(pool)
            GitReaper.add_wait
            GitReaper.execute "git commit -m \" commit #{@@commits} to pool[#{pool}] at #{Time.now.strftime("%H:%M - %d/%m/%Y")} \""
    end

    def self.atomic(why, pool)
        open('why_commit.txt', 'a') do |file|
            file.puts "#{Time.now.strftime("%d/%m/%Y %H:%M")}:pool[#{pool}]: #{why}"
        end
        changes = why.strip.split(",")
        changes.map! {|item| item = "* #{item.strip}"}
        
        open('pull_me.txt', 'a') do |file|
            file.puts "### pool[#{pool}]:"
            file.puts changes
            file.puts
        end
        GitReaper.add_wait
        GitReaper.execute "git commit -m \"pool[#{pool}]: #{why}\""
        
    end

    def self.exit(exit_type, pool, branch)
        case exit_type
        when "push"
            puts "Summarize changes made:"
            summary = gets.chomp
            GitReaper.atomic(summary, pool)
            GitReaper.start
        when "kill"
            puts "Wiping commits and exiting"
            system "git reset HEAD~"
        when "reap"
            puts "Summarize final changes:"
            summary = gets.chomp
            GitReaper.atomic(summary, pool)
            puts "Reaping #{@@commits-1} commits to pool on branch: #{branch}"
            GitReaper.execute "git push -u origin #{branch}"
        else
            puts "Returning to loop"
            GitReaper.threader(branch)
        end
    end

    def self.threader(branch)
       thread_pool = []
       thread_fork = [0,1]
       thread_bits = []
       thread_bits = Threader.bits
       thread_pool.push(Threader.bits_adjs[rand(Threader.bits_adjs.length)] + "-")
       thread_pool.push(Threader.bits_verbs[rand(Threader.bits_verbs.length)] + "-")
       thread_pool.push(Threader.bits_nouns[rand(Threader.bits_nouns.length)] + "-")

        6.times do
            do_fork = thread_fork[rand(thread_fork.length)]
            if do_fork == 0
                thread_pool.push(rand(9))
            else
                thread_pool.push(thread_bits[rand(thread_bits.length)])
            end
        end
        puts "Preparing to Reap on #{branch} branch."
        reaper = Thread.new do
            
            while true
                GitReaper.commit_loop(thread_pool.join(''))
            end
            
        end
        
        gets
        reaper.kill
        puts "How do you wish to exit?"
        puts "'push': pushes all commits to branch\n'kill': wipes commits and exits program\n'reap': pushes all changes"
        exit_type = gets.chomp
        GitReaper.exit(exit_type, thread_pool.join(''), branch)
        
        
    end

    def self.start

        @@pushes > 0 ? @@pushes += 1 : open('pull_me.txt', 'w') {|f| f.puts ""}; @@pushes += 1
        puts "Branch to push?"
        branch = gets.chomp
        GitReaper.threader(branch)
    end

    

end

GitReaper.start


httpdec.rb
#!/usr/bin/env ruby


class GitDeclare
    @@pushes = 0
    @@stage = 0
    @@changes = []
    @@color_red = "\033[31m"
    @@color_green = "\033[32m"
    @@color_default = "\033[0m"
    @@commits = 1
    @@starttime = Time.now.strftime("%H:%M")
    @@date = Time.now.strftime("%d/%m/%Y")
    @@time = nil
    @@pool = nil
    @@branch = nil
    @@declare = nil
    @@uri = "http://localhost:3000/declarations"

    def initialize; end

    def self.current_time
        Time.now.strftime("%H:%M")
    end

    def self.execute(param)
        stalker = %x{#{param}}
        
        if stalker.include? "nothing to commit" 
            
        elsif stalker.include? "insert"
            puts @@color_green + stalker + @@color_default
            puts "#{@@commits} changes saved"
            @@commits += 1
        end
    end

    def self.add_wait
        sleep 1
        GitDeclare.execute "git add ."
        sleep 1
    end

    def self.commit_loop(pool)
            GitDeclare.add_wait
            GitDeclare.execute "git commit -m \" #{pool} \""
    end

    def self.atomic(summary, pool)
        open("#{Dir.pwd}/changelog.txt", 'a') do |file|
            file.puts "#{@@date}: #{@@time} - #{GitDeclare.current_time}:pool[#{pool}]"
        end
        open("#{Dir.pwd}/readme.md", 'a') do |file|
            file.puts "\n##### #{@@date}: #{@@time} - #{GitDeclare.current_time}:pool[#{pool}]"
        end
        GitDeclare.post("#{@@uri}/#{@@declare}/entries", body: {content: pool})
        @@changes << pool
        if @@stage == 1
            GitDeclare.put("#{@@uri}/#{@@declare}", body: {content: summary})
            @@changes.map! {|item| item = "* #{item.strip}"}
            open('pull_me.txt', 'a') do |file|
                file.puts "[#{summary}]"
                file.puts "### #{@@date}[#{@@starttime} - #{GitDeclare.current_time}]:"
                file.puts @@changes
                file.puts
            end
        end
        GitDeclare.add_wait
        GitDeclare.execute "git commit -m \"#{pool}\""
        
    end

    def self.exit(exit_type, pool, branch)
        case exit_type
        when "new"
            GitDeclare.atomic(nil, pool)
            GitDeclare.start
        when "reset"
            puts "Wiping commits and exiting"
            system "git reset HEAD~"
        when "push"
            puts "Summarize final changes:"
            summary = gets.chomp
            @@stage = 1
            GitDeclare.atomic(summary, pool)
            GitDeclare.execute "git push -u origin #{branch}"
        when "switch"
            GitDeclare.atomic(nil, pool)
            puts "What branch are you working on?"; @@branch = gets.chomp
            GitDeclare.start
        else
            puts "Returning to loop"
            GitDeclare.threader(branch)
        end
    end

    def self.threader(branch)
        puts "What are you working on with the #{branch} branch?"
        @@pool = gets.chomp
        puts "You're now working on: \"#{@@pool}\" on #{branch} branch. #{@@color_red}GitDeclare is watching for changes#{@@color_default}."
        puts "When you're done with this change, press [Enter] to make a commit and start a new declaration."
        declare = Thread.new do
            
            while true
                GitDeclare.commit_loop(@@pool)
            end
            
        end
        
        gets
        declare.kill
        puts "How do you wish to exit?"
        puts "
        'new': logs the commit pool and starts a new declaration
        'reset': wipes commits and exits program
        'push': pushes all changes
        'switch': logs commit, starts new declare, and switches branch"
        exit_type = gets.chomp
        GitDeclare.exit(exit_type, @@pool, branch)
        
        
    end

    def self.start
        @@time = GitDeclare.current_time
        x = %x(git rev-parse --abbrev-ref HEAD)
        @@branch = x.strip
        puts "On #{@@branch} branch"
        if @@pushes > 0
            @@pushes += 1
        else
            open('pull_me.txt', 'w') {|f| f.puts ""}
            @@pushes += 1
            res = GitDeclare.post(@@uri, body: {content: "New Declaration"})
            body = JSON.parse(res.body)
            @@declare = body["id"]
            
        end
        if @@branch == "master" then puts "What branch are you working on?"; @@branch = gets.chomp end
        
        GitDeclare.threader(@@branch)
    end

    

end

GitDeclare.start


sandbox.rb
#!/usr/bin/env ruby

require 'httparty'

class GitDeclare
    include HTTParty
    @@pushes = 0
    @@stage = 0
    @@changes = []
    @@color_red = "\033[31m"
    @@color_green = "\033[32m"
    @@color_default = "\033[0m"
    @@commits = 1
    @@starttime = Time.now.strftime("%H:%M")
    @@date = Time.now.strftime("%d/%m/%Y")
    @@time = nil
    @@pool = nil
    @@branch = nil
    @@declare = nil
    @@uri = "http://localhost:3000/declarations"
    @@cwd = nil
    @@origin = nil

    def initialize; end

    def self.current_time
        Time.now.strftime("%H:%M")
    end

    def self.execute(param)
        puts param
    end

    def self.add_wait
        sleep 1
        GitDeclare.execute "git add ."
        sleep 1
    end

    def self.commit_loop(pool)
            GitDeclare.add_wait
            GitDeclare.execute "git commit -m \" #{pool} \""
    end

    def self.atomic(summary, pool)
        open("#{Dir.pwd}/changelog.txt", 'a') do |file|
            file.puts "#{@@date}: #{@@time} - #{GitDeclare.current_time}:pool[#{pool}]"
        end
        open("#{Dir.pwd}/readme.md", 'a') do |file|
            file.puts "\n##### #{@@date}: #{@@time} - #{GitDeclare.current_time}:pool[#{pool}]"
        end
        GitDeclare.post("#{@@uri}/#{@@declare}/entries", body: {content: pool})
        @@changes << pool
        if @@stage == 1
            GitDeclare.put("#{@@uri}/#{@@declare}", body: {content: summary, directory: @@cwd, git: @@origin})
            @@changes.map! {|item| item = "* #{item.strip}"}
            open('pull_me.txt', 'a') do |file|
                file.puts "[#{summary}]"
                file.puts "### #{@@date}[#{@@starttime} - #{GitDeclare.current_time}]:"
                file.puts @@changes
                file.puts
            end
        end
        GitDeclare.add_wait
        GitDeclare.execute "git commit -m \"#{pool}\""
        
    end

    def self.exit(exit_type, pool, branch)
        case exit_type
        when "new"
            GitDeclare.atomic(nil, pool)
            GitDeclare.start
        when "reset"
            puts "Wiping commits and exiting"
            system "git reset HEAD~"
        when "push"
            puts "Summarize final changes:"
            summary = gets.chomp
            @@stage = 1
            GitDeclare.atomic(summary, pool)
            GitDeclare.execute "git push -u origin #{branch}"
        when "switch"
            GitDeclare.atomic(nil, pool)
            puts "What branch are you working on?"; @@branch = gets.chomp
            GitDeclare.start
        else
            puts "Returning to loop"
            GitDeclare.threader(branch)
        end
    end

    def self.threader(branch)
        puts "What are you working on with the #{branch} branch?"
        @@pool = gets.chomp
        puts "You're now working on: \"#{@@pool}\" on #{branch} branch. #{@@color_red}GitDeclare is watching for changes#{@@color_default}."
        puts "When you're done with this change, press [Enter] to make a commit and start a new declaration."
        declare = Thread.new do
            
            while true
                GitDeclare.commit_loop(@@pool)
            end
            
        end
        
        gets
        declare.kill
        puts "How do you wish to exit?"
        puts "
        'new': logs the commit pool and starts a new declaration
        'reset': wipes commits and exits program
        'push': pushes all changes
        'switch': logs commit, starts new declare, and switches branch"
        exit_type = gets.chomp
        GitDeclare.exit(exit_type, @@pool, branch)
        
        
    end

    def self.start
        
        @@time = GitDeclare.current_time
        @@cwd = Dir.pwd.split("/").last
        @@origin = %x(git remote get-url --push origin)
        @@branch = %x(git rev-parse --abbrev-ref HEAD).strip
        puts "On #{@@branch} branch"
        if @@pushes > 0
            @@pushes += 1
        else
            open('pull_me.txt', 'w') {|f| f.puts ""}
            @@pushes += 1
            res = GitDeclare.post(@@uri, body: {content: "New Declaration", directory: "New", git: "New"})
            body = JSON.parse(res.body)
            @@declare = body["id"]
            
        end
        if @@branch == "master" then puts "What branch are you working on?"; @@branch = gets.chomp end
        
        GitDeclare.threader(@@branch)
    end

    

end

GitDeclare.start


gitdeclare.rb



class GitDeclare
    @@pushes = 0
    @@stage = 0
    @@changes = []
    @@color_red = "\033[31m"
    @@color_green = "\033[32m"
    @@color_default = "\033[0m"
    @@commits = 1
    @@starttime = Time.now.strftime("%H:%M")
    @@date = Time.now.strftime("%d/%m/%Y")
    @@time = nil
    @@pool = nil
    @@branch = nil

    def initialize; end

    def self.current_time
        Time.now.strftime("%H:%M")
    end

    def self.execute(param)
        stalker = %x{#{param}}
        
        if stalker.include? "nothing to commit" 
            
        elsif stalker.include? "insert"
            puts @@color_green + stalker + @@color_default
            puts "#{@@commits} changes saved"
            @@commits += 1
        end
    end

    def self.add_wait
        sleep 1
        GitDeclare.execute "git add ."
        sleep 1
    end

    def self.commit_loop(pool)
            GitDeclare.add_wait
            GitDeclare.execute "git commit -m \" #{pool} \""
    end

    def self.atomic(summary, pool, goal=nil)
        open("#{Dir.pwd}/changelog.txt", 'a') do |file|
            file.puts "#{@@date}: #{@@time} - #{GitDeclare.current_time}:pool[#{pool}]"
        end
        
        @@changes << pool
        if @@stage == 1 && @@branch != "master"
            @@changes.map! {|item| item = "* #{item.strip}"}
            open('pull_me.txt', 'a') do |file|
                file.puts "[#{summary}]"
                file.puts "### #{@@date}[#{@@starttime} - #{GitDeclare.current_time}]:"
                file.puts @@changes
                file.puts
                file.puts "What's next? #{goal}"
            end
        end
        GitDeclare.add_wait
        GitDeclare.execute "git commit -m \"#{pool}\""
        
    end

    def self.exit(exit_type, pool, branch)
        case exit_type
        when "new"
            GitDeclare.atomic(nil, pool)
            GitDeclare.start
        when "reset"
            puts "Wiping commits and exiting"
            system "git reset HEAD~"
        when "push"
            if @@branch != "master"
                puts "Summarize final changes:"
                summary = gets.chomp
                puts "What is your next goal?"
                goal = gets.chomp
                @@stage = 1
                GitDeclare.atomic(summary, pool, goal)
            end
            GitDeclare.execute "git push -u origin #{branch}"
        when "switch"
            GitDeclare.atomic(nil, pool)
            puts "What branch are you working on?"; @@branch = gets.chomp
            GitDeclare.start
        else
            puts "Returning to loop"
            GitDeclare.threader(branch)
        end
    end

    def self.threader(branch)
        puts "What are you working on with the #{branch} branch?"
        @@pool = gets.chomp
        puts "You're now working on: \"#{@@pool}\" on #{branch} branch. #{@@color_red}GitDeclare is watching for changes#{@@color_default}."
        puts "When you're done with this change, press [Enter] to make a commit and start a new declaration."
        declare = Thread.new do
            
            while true
                GitDeclare.commit_loop(@@pool)
            end
            
        end
        
        gets
        declare.kill
        puts "How do you wish to exit?"
        puts "
        'new': logs the commit pool and starts a new declaration
        'reset': wipes commits and exits program
        'push': pushes all changes
        'switch': logs commit, starts new declare, and switches branch"
        exit_type = gets.chomp
        GitDeclare.exit(exit_type, @@pool, branch)
        
        
    end

    def self.start
        @@time = GitDeclare.current_time
        x = %x(git rev-parse --abbrev-ref HEAD)
        @@branch = x.strip
        puts "On #{@@branch} branch"
        @@pushes > 0 ? @@pushes += 1 : open('pull_me.txt', 'w') {|f| f.puts ""}; @@pushes += 1
        if @@branch == nil then puts "What branch are you working on?"; @@branch = gets.chomp end
        
        GitDeclare.threader(@@branch)
    end

    

end

GitDeclare.start


gdec_http.rb
#!/usr/bin/env ruby

require 'httparty'

class GitDeclare
    include HTTParty
    @@pushes = 0
    @@stage = 0
    @@changes = []
    @@color_red = "\033[31m"
    @@color_green = "\033[32m"
    @@color_default = "\033[0m"
    @@commits = 1
    @@starttime = Time.now.strftime("%H:%M")
    @@date = Time.now.strftime("%d/%m/%Y")
    @@time = nil
    @@pool = nil
    @@branch = nil
    @@declare = nil
    @@uri = "http://localhost:3000/declarations"
    @@cwd = nil
    @@origin = nil

    def initialize; end

    def self.current_time
        Time.now.strftime("%H:%M")
    end

    def self.execute(param)
        stalker = %x{#{param}}
        
        if stalker.include? "nothing to commit" 
            
        elsif stalker.include? "insert"
            puts @@color_green + stalker + @@color_default
            puts "#{@@commits} changes saved"
            @@commits += 1
        end
    end

    def self.add_wait
        sleep 1
        GitDeclare.execute "git add ."
        sleep 1
    end

    def self.commit_loop(pool)
            GitDeclare.add_wait
            GitDeclare.execute "git commit -m \" #{pool} \""
    end

    def self.atomic(summary, pool, goal=nil)
        open("#{Dir.pwd}/changelog.txt", 'a') do |file|
            file.puts "#{@@date}: #{@@time} - #{GitDeclare.current_time}:pool[#{pool}]"
        end

        GitDeclare.post("#{@@uri}/#{@@declare}/entries", body: {content: pool})
        @@changes << pool
        if @@stage == 1 && @@branch != "master"
            GitDeclare.put("#{@@uri}/#{@@declare}", body: {content: summary, directory: @@cwd, git: @@origin})
            @@changes.map! {|item| item = "* #{item.strip}"}
            open('pull_me.txt', 'a') do |file|
                file.puts "[#{summary}]"
                file.puts "### #{@@date}[#{@@starttime} - #{GitDeclare.current_time}]:"
                file.puts @@changes
                file.puts
                file.puts "What's next? #{goal}"
            end
        end
        GitDeclare.add_wait
        GitDeclare.execute "git commit -m \"#{pool}\""
        
    end

    def self.exit(exit_type, pool, branch)
        case exit_type
        when "new"
            GitDeclare.atomic(nil, pool)
            GitDeclare.start
        when "reset"
            puts "Wiping commits and exiting"
            system "git reset HEAD~"
        when "push"
            if @@branch != "master"
                puts "Summarize final changes:"
                summary = gets.chomp
                puts "What is your next goal?"
                goal = gets.chomp
                @@stage = 1
                GitDeclare.atomic(summary, pool, goal)
            end
            GitDeclare.execute "git push -u origin #{branch}"
        when "switch"
            GitDeclare.atomic(nil, pool)
            puts "What branch are you working on?"; @@branch = gets.chomp
            GitDeclare.start
        else
            puts "Returning to loop"
            GitDeclare.threader(branch)
        end
    end

    def self.threader(branch)
        puts "What are you working on with the #{branch} branch?"
        @@pool = gets.chomp
        puts "You're now working on: \"#{@@pool}\" on #{branch} branch. #{@@color_red}GitDeclare is watching for changes#{@@color_default}."
        puts "When you're done with this change, press [Enter] to make a commit and start a new declaration."
        declare = Thread.new do
            
            while true
                GitDeclare.commit_loop(@@pool)
            end
            
        end
        
        gets
        declare.kill
        puts "How do you wish to exit?"
        puts "
        'new': logs the commit pool and starts a new declaration
        'reset': wipes commits and exits program
        'push': pushes all changes
        'switch': logs commit, starts new declare, and switches branch"
        exit_type = gets.chomp
        GitDeclare.exit(exit_type, @@pool, branch)
        
        
    end

    def self.start
        
        @@time = GitDeclare.current_time
        @@cwd = Dir.pwd.split("/").last
        @@origin = %x(git remote get-url --push origin)
        @@branch = %x(git rev-parse --abbrev-ref HEAD).strip
        puts "On #{@@branch} branch"
        if @@pushes > 0
            @@pushes += 1
        else
            open('pull_me.txt', 'w') {|f| f.puts ""}
            @@pushes += 1
            res = GitDeclare.post(@@uri, body: {content: "Working on...", directory: "#{@@cwd}", git: "#"})
            body = JSON.parse(res.body)
            @@declare = body["id"]
            
        end
        if @@branch == "master" then puts "What branch are you working on?"; @@branch = gets.chomp end
        
        GitDeclare.threader(@@branch)
    end

    

end

GitDeclare.start
